### PR TITLE
[Training] Add Qwen3-Omni multimodal RL support

### PR DESCRIPTION
**Pull request by CalvinXKY**
## Summary

- Add Megatron Bridge + preprocessing for Qwen3-Omni Thinker (vision/audio, DeepStack, M-RoPE)
- Extend rollout render path for `audio_url` / `video_url`, plus optional encoder-feature dump matching for train/infer alignment
- Ship temporary vLLM compatibility changes in `docker/patch/latest/vllm.patch` (audio `disable_tp`, feature dumps) and model args script

Expert weight mappings follow the same per-expert HF layout as `glm4v_moe` (`GatedMLPMapping` / `AutoMapping` on `weight*`), so bridge load and live `save_hf` / weight sync emit standard `experts.{e}.gate_proj|up_proj|down_proj` keys (no fused `gate_up_proj` workaround).

This PR is **self-contained**: the audio `disable_tp` fix needed for TP>1 is already in `docker/patch/latest/vllm.patch`, so Vime does **not** depend on the upstream vLLM change to land first.

Related upstream tracking: https://github.com/vllm-project/vllm/pull/50858  
Once that PR is merged into vLLM and Vime's base image picks it up, the corresponding `disable_tp` hunk in `vllm.patch` can be removed (feature-dump hunks may still remain until a longer-term shared-feature path exists).

## Known limitation (offline torch_dist → HF)

Live bridge export (`save_hf` / weight sync after loading the Megatron model) uses the mappings above.

Directly reading a torch_dist DCP flat state dict (keys without `layers.N`, and `mlp.experts.experts.*` 4D tensors) is a different path. If you need that offline conversion today, use the temporary script posted in the PR comments (validated on AVQA checkpoints). Prefer loading the Megatron model then exporting when possible.

## Test plan

Verified on 8x A800, Qwen3-Omni-30B-A3B-Instruct, AVQA (image+audio), TP=4 / EP=2 / colocate:

- [x] Build / use an image with `docker/patch/latest/vllm.patch` applied (`disable_tp` + TP-rank-0 feature dumps)
- [x] Launch Qwen3-Omni-30B-A3B Thinker with `--megatron-to-hf-mode bridge` (load + weight sync OK)
- [x] Confirm TP>1 rollout no longer fails on audio encoder head divisibility (TP=4)
- [x] Confirm audio/video samples go through `/v1/chat/completions/render`; dumps only on TP rank 0
- [x] Short GRPO/colocate smoke: `train_rollout_logprob_abs_diff` mean **0.055** (10 steps)
- [x] Real AVQA reward training **400 steps**: logprob abs diff stable ~0.02–0.08; Match Acc **25% → 29%** at iter_299 (external vLLM eval, convert via comment script)

### Smoke metrics (10 steps)

| Metric | Value |
|--------|-------|
| mean logprob abs diff | 0.0552 |
| min / max | 0.0465 / 0.0689 |
| training stability | no NaN / no divergence |

### Longer run (400 steps, AVQA real reward)

| Metric | Value |
|--------|-------|
| logprob abs diff | ~0.042 mean over steps 300–399 |
| Match Accuracy (baseline → iter_299) | 25% → 29% (+4pp) |
| Token Recall | 0.340 → 0.393 |

### CODE REVIEW

**Comment by CalvinXKY** [comment]
## Smoke test results (2026-08-03)

Environment: 8x A800, container `vime-omni`, Qwen3-Omni-30B-A3B-Instruct, AVQA image+audio, TP=4 / EP=2 / colocate.

| Check | Result |
|-------|--------|
| `vllm.patch` applied (`disable_tp` + dump TP guard) | PASS |
| Bridge load + weight sync | PASS |
| TP>1 audio encoder (no 20%TP error) | PASS |
| multimodal `/render` + rank-0 dumps | PASS |
| train/infer `logprob_abs_diff` (10 steps) | **PASS** — mean 0.0552, range 0.0465–0.0689 |
| `tests/test_vllm_rollout.py` | PARTIAL — 30/33; 3 failures appear test-env/tokenizer related |

**Comment by CalvinXKY** [comment]
## Follow-up notes from 400-step AVQA training

### What landed in this update
- Removed `examples/qwen3_omni` from the PR (keep model args in `scripts/models/qwen3_omni_moe.sh` only).
- Bridge expert mappings now use **per-expert** HF keys (`GatedMLPMapping` / `AutoMapping` on `weight*`), aligned with `glm4v_moe`, instead of fused `gate_up_proj` + `maybe_modify_loaded_hf_weight`.

### Offline torch_dist → HF (temporary)
Live path: load Megatron model from checkpoint → `save_hf` / weight sync (preferred).

If you must convert a **torch_dist DCP** directory by reading flat keys directly (no Megatron reload), the DCP layout uses:
- flat layer dim, e.g. `language_model.decoder.layers.self_attention.linear_qkv.weight` → `[num_layers, …]`
- experts as `…mlp.experts.experts.linear_fc1.weight` → `[num_layers, num_experts, 2*inter, hidden]`
- QKV packed with **TP-interleaved** `[Q0,K0,V0, Q1,K1,V1, …]` after gathering

Temporary workaround script (parameterize paths / TP before use):

```python
"""Temporary offline torch_dist -> HF for Qwen3-Omni-MoE (not part of this PR)."""
import glob
import json
import os
import shutil

import torch
import torch.distributed.checkpoint as dcp
from safetensors import safe_open
from safetensors.torch import save_file
from torch.distributed.checkpoint import FileSystemReader

# --- configure me ---
CKPT_DIR = "/path/to/iter_0000299"
ORIG_HF_DIR = "/path/to/Qwen3-Omni-30B-A3B-Instruct"
OUTPUT_DIR = "/path/to/output_hf"
NUM_LAYERS = 48
NUM_EXPERTS = 128
MOE_INTER = 768
NUM_Q_HEADS = 32
NUM_KV_GROUPS = 4
HEAD_DIM = 128
TP_SIZE = 4  # must match the training TP used to write the checkpoint
Q_DIM = NUM_Q_HEADS * HEAD_DIM
KV_DIM = NUM_KV_GROUPS * HEAD_DIM
Q_PER_RANK = Q_DIM // TP_SIZE
KV_PER_RANK = KV_DIM // TP_SIZE
QKV_CHUNK = Q_PER_RANK + 2 * KV_PER_RANK


def load_torch_dist_state(ckpt_dir):
    reader = FileSystemReader(ckpt_dir)
    metadata = reader.read_metadata()
    state_dict = {}
    for key, storage_meta in metadata.state_dict_metadata.items():
        if hasattr(storage_meta, "size") and storage_meta.size:
            state_dict[key] = torch.empty(list(storage_meta.size), dtype=storage_meta.properties.dtype)
    dcp.load(state_dict, storage_reader=reader)
    return state_dict


def load_hf_state_dict(hf_dir):
    state_dict = {}
    for f in sorted(glob.glob(os.path.join(hf_dir, "*.safetensors"))):
        with safe_open(f, framework="pt") as st:
            for k in st.keys():
                state_dict[k] = st.get_tensor(k)
    return state_dict


def convert_torch_dist_to_hf(td_state, hf_state):
    if "language_model.embedding.word_embeddings.weight" in td_state:
        hf_state["thinker.model.embed_tokens.weight"] = td_state[
            "language_model.embedding.word_embeddings.weight"
        ].clone()
    if "language_model.output_layer.weight" in td_state:
        hf_state["thinker.lm_head.weight"] = td_state["language_model.output_layer.weight"].clone()
    if "language_model.decoder.final_layernorm.weight" in td_state:
        hf_state["thinker.model.norm.weight"] = td_state[
            "language_model.decoder.final_layernorm.weight"
        ].clone()

    qkv_w = td_state.get("language_model.decoder.layers.self_attention.linear_qkv.weight")
    qkv_ln = td_state.get("language_model.decoder.layers.self_attention.linear_qkv.layer_norm_weight")
    q_ln = td_state.get("language_model.decoder.layers.self_attention.q_layernorm.weight")
    k_ln = td_state.get("language_model.decoder.layers.self_attention.k_layernorm.weight")
    proj_w = td_state.get("language_model.decoder.layers.self_attention.linear_proj.weight")
    pre_mlp_ln = td_state.get("language_model.decoder.layers.pre_mlp_layernorm.weight")
    router_w = td_state.get("language_model.decoder.layers.mlp.router.weight")
    expert_fc1 = td_state.get("language_model.decoder.layers.mlp.experts.experts.linear_fc1.weight")
    expert_fc2 = td_state.get("language_model.decoder.layers.mlp.experts.experts.linear_fc2.weight")

    for layer in range(NUM_LAYERS):
        prefix = f"thinker.model.layers.{layer}"
        if qkv_w is not None:
            qkv = qkv_w[layer]
            q_parts, k_parts, v_parts = [], [], []
            for r in range(TP_SIZE):
                start = r * QKV_CHUNK
                q_parts.append(qkv[start : start + Q_PER_RANK])
                k_parts.append(qkv[start + Q_PER_RANK : start + Q_PER_RANK + KV_PER_RANK])
                v_parts.append(qkv[start + Q_PER_RANK + KV_PER_RANK : start + QKV_CHUNK])
            hf_state[f"{prefix}.self_attn.q_proj.weight"] = torch.cat(q_parts).clone()
            hf_state[f"{prefix}.self_attn.k_proj.weight"] = torch.cat(k_parts).clone()
            hf_state[f"{prefix}.self_attn.v_proj.weight"] = torch.cat(v_parts).clone()
        if qkv_ln is not None:
            hf_state[f"{prefix}.input_layernorm.weight"] = qkv_ln[layer].clone()
        if q_ln is not None:
            hf_state[f"{prefix}.self_attn.q_norm.weight"] = q_ln[layer].clone()
        if k_ln is not None:
            hf_state[f"{prefix}.self_attn.k_norm.weight"] = k_ln[layer].clone()
        if proj_w is not None:
            hf_state[f"{prefix}.self_attn.o_proj.weight"] = proj_w[layer].clone()
        if pre_mlp_ln is not None:
            hf_state[f"{prefix}.post_attention_layernorm.weight"] = pre_mlp_ln[layer].clone()
        if router_w is not None:
            hf_state[f"{prefix}.mlp.gate.weight"] = router_w[layer].clone()
        if expert_fc1 is not None:
            for e in range(NUM_EXPERTS):
                fc1 = expert_fc1[layer, e]
                hf_state[f"{prefix}.mlp.experts.{e}.gate_proj.weight"] = fc1[:MOE_INTER].clone()
                hf_state[f"{prefix}.mlp.experts.{e}.up_proj.weight"] = fc1[MOE_INTER:].clone()
        if expert_fc2 is not None:
            for e in range(NUM_EXPERTS):
                hf_state[f"{prefix}.mlp.experts.{e}.down_proj.weight"] = expert_fc2[layer, e].clone()
    return hf_state


def save_hf_checkpoint(hf_state, output_dir, orig_hf_dir):
    os.makedirs(output_dir, exist_ok=True)
    for f in os.listdir(orig_hf_dir):
        if not f.endswith((".safetensors", ".bin")):
            src = os.path.join(orig_hf_dir, f)
            if os.path.isfile(src):
                shutil.copy2(src, os.path.join(output_dir, f))
    max_shard = 5 * 1024**3
    shard, size, idx = {}, 0, 1
    for key in sorted(hf_state):
        t = hf_state[key]
        nbytes = t.numel() * t.element_size()
        if size + nbytes > max_shard and shard:
            save_file(shard, os.path.join(output_dir, f"model-{idx:05d}-of-99999.safetensors"))
            shard, size, idx = {}, 0, idx + 1
        shard[key] = t.contiguous()
        size += nbytes
    if shard:
        save_file(shard, os.path.join(output_dir, f"model-{idx:05d}-of-99999.safetensors"))
    total = idx
    weight_map = {}
    for i in range(1, total + 1):
        old = os.path.join(output_dir, f"model-{i:05d}-of-99999.safetensors")
        new_name = f"model-{i:05d}-of-{total:05d}.safetensors"
        os.rename(old, os.path.join(output_dir, new_name))
        with safe_open(os.path.join(output_dir, new_name), framework="pt") as st:
            for k in st.keys():
                weight_map[k] = new_name
    with open(os.path.join(output_dir, "model.safetensors.index.json"), "w") as f:
        json.dump(
            {
                "metadata": {"total_size": sum(v.numel() * v.element_size() for v in hf_state.values())},
                "weight_map": weight_map,
            },
            f,
            indent=2,
        )


if __name__ == "__main__":
    hf = load_hf_state_dict(ORIG_HF_DIR)
    td = load_torch_dist_state(CKPT_DIR)
    hf = convert_torch_dist_to_hf(td, hf)
    save_hf_checkpoint(hf, OUTPUT_DIR, ORIG_HF_DIR)
    print(f"Wrote {len(hf)} keys to {OUTPUT_DIR}")
```

Validated on AVQA ckpts: ~28010 HF keys; expert / QKV cos_sim ≈ 1.0 vs base weights after correct TP deinterleave.

### Large MoE checkpoint save tip
30B MoE DCP is ~400GB; default NCCL timeout (10 min) can fail at barrier after data write. Prefer raising `NCCL_TIMEOUT` / `NCCL_TIMEOUT_MS`, and verify `.metadata` + `metadata.json` exist after save. Incomplete metadata can sometimes be recovered by copying those two files from a sibling complete iteration (same parallel layout).

**Comment by CalvinXKY** [comment]
Qwen3-Omni AVQA Long-Run Test

Goal: Run 400-step GRPO with real AVQA reward on Qwen3-Omni-30B, and check (1) train/infer consistency stays stable, (2) held-out AVQA score improves.

Setup: vime + Megatron bridge + vLLM, 8×A800 colocate (TP=4, EP=2), 500 train / 100 eval samples. Eval checkpoint: iter_299 (300 steps).

image

image

**Comment by aoshen02** [comment]
Just did some refactor to align with the new code, could you please retest it to see if it works? I'm ok with the code change right now.

**Comment by CalvinXKY** [comment]
> Just did some refactor to align with the new code, could you please retest it to see if it works? I'm ok with the code change right now.

Great! I'll test it e2e.

**Comment by CalvinXKY** [comment]
**Result: PASS — full training pipeline completed successfully**

 Re-verified on 2026-08-29 with `vllm/vime:cu13-latest` image (vLLM 0.25.1) on A800-server2 (8×A800 80GB), using Qwen3-Omni-30B-A3B-Instruct model and AVQA-R1-6K dataset.

  Test configuration:
  - TP=2, EP=4, SP=True, colocate mode, --no-offload-train
  - vLLM rollout: TP=8, EP=8
  - 5-step GRPO, rollout-batch-size=2, n-samples-per-prompt=2

**Review by aoshen02 — APPROVED** [review]

**Review by Meihan-chen — APPROVED** [review]
